builtin/docker/kaniko.go

```diff
@@ -0,0 +1,148 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+
+	"github.com/docker/distribution/reference"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"github.com/hashicorp/waypoint/internal/assets"
+	"github.com/hashicorp/waypoint/internal/ociregistry"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// HasImg returns true if "img" is available on the PATH.
+//
+// This doesn't do any fancy checking that "img" is the "img" we expect.
+// We can make the checking here more advanced later.
+func HasKaniko() bool {
+	_, err := os.Stat("/kaniko/executor")
+	return err == nil
+}
+
+func (b *Builder) buildWithKaniko(
+	ctx context.Context,
+	ui terminal.UI,
+	sg terminal.StepGroup,
+	log hclog.Logger,
+	dockerfilePath string,
+	contextDir string,
+	buildArgs map[string]*string,
+	ai *AccessInfo,
+) (*Image, error) {
+	step := sg.Add("Building Docker image with kaniko...")
+	defer func() {
+		if step != nil {
+			step.Abort()
+		}
+	}()
+
+	target := &Image{
+		Image:    ai.Image,
+		Tag:      ai.Tag,
+		Location: &Image_Registry{Registry: &empty.Empty{}},
+	}
+
+	var auth string
+
+	if ai.Auth != nil {
+		switch sv := ai.Auth.(type) {
+		case *AccessInfo_Encoded:
+			user, pass, err := CredentialsFromConfig(sv.Encoded)
+			if err != nil {
+				return nil, err
+			}
+			auth = ociregistry.BasicAuth(user, pass)
+		case *AccessInfo_Header:
+			auth = sv.Header
+		}
+	}
+
+	// Determine the host that we're setting auth for. We have to parse the
+	// image for this cause it may not contain a host. Luckily Docker has
+	// libs to normalize this all for us.
+	log.Trace("determining host for auth configuration", "image", target.Name())
+	ref, err := reference.ParseNormalizedNamed(target.Name())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to parse image name: %s", err)
+	}
+	host := reference.Domain(ref)
+	log.Trace("auth host", "host", host)
+
+	var os ociregistry.Server
+	os.DisableEntrypoint = b.config.DisableCEB
+	os.Auth = auth
+	os.Logger = log
+	os.Upstream = "http://" + host
```

Comment:
Fix this http to https unless host is localhost OR they set configuration for insecure.